taichi/analysis/build_cfg.cpp

```diff
@@ -0,0 +1,184 @@
+#include "taichi/ir/cfg.h"
+#include "taichi/ir/ir.h"
+#include "taichi/ir/statements.h"
+
+TLANG_NAMESPACE_BEGIN
+
+// Build a control-flow graph
+class CFGBuilder : public IRVisitor {
+ private:
+  std::unique_ptr<ControlFlowGraph> graph;
+  Block *current_block;
+  CFGNode *last_node_in_current_block;
+  std::vector<CFGNode *> continues_in_current_loop;
+  std::vector<CFGNode *> breaks_in_current_loop;
+  int current_stmt_id;
+  int begin_location;
+  std::vector<CFGNode *> prev_nodes;
+
+ public:
+  CFGBuilder()
+      : current_block(nullptr),
+        last_node_in_current_block(nullptr),
+        current_stmt_id(-1),
+        begin_location(-1) {
+    allow_undefined_visitor = true;
+    invoke_default_visitor = true;
+    graph = std::make_unique<ControlFlowGraph>();
+  }
+
+  void visit(Stmt *stmt) override {
+    if (stmt->is_container_statement()) {
+      TI_ERROR("Visitor for container stmt undefined.");
+    }
+  }
+
+  CFGNode *new_node(int next_begin_location) {
+    auto node =
+        graph->push_back(current_block, begin_location, current_stmt_id - 1,
+                         last_node_in_current_block);
+    for (auto &prev_node : prev_nodes) {
+      CFGNode::add_edge(prev_node, node);
+    }
+    prev_nodes.clear();
+    begin_location = next_begin_location;
+    last_node_in_current_block = node;
+    return node;
+  }
+
+  void visit(ContinueStmt *stmt) override {
+    // Don't put ContinueStmt in any nodes.
+    continues_in_current_loop.push_back(new_node(current_stmt_id + 1));
```

Comment:
Slightly confused about this: what is this `new_node` for, if not `continue`?

Reply:
For example,
```
a = 1
b = 1
continue
c = 1
```
`new_node` will construct a node with
```
a = 1
b = 1
```
and the next node will begin at `current_stmt_id + 1`, that is,
```
c = 1
```